ExampleMod/Content/TileEntities/BasicTileEntity.cs

```diff
@@ -0,0 +1,239 @@
+			TileObjectData.newTile.StyleHorizontal = true;
+
+			TileObjectData.addTile(Type);
+
+			// Etc
+			AddMapEntry(new Color(200, 200, 200), CreateMapEntryName(), MapHoverText);
+
+			StatusText = this.GetLocalization(nameof(StatusText));
+		}
+
+		public override void KillMultiTile(int i, int j, int frameX, int frameY) {
+			// When the tile is removed, we need to remove the Tile Entity as well.
+			ModContent.GetInstance<BasicTileEntity>().Kill(i, j);
+		}
+
+		// The following hooks all show accessing the Tile Entity and using it to adjust the behavior and look of this Tile.
+
+		public static string MapHoverText(string name, int i, int j) {
+			if (TryGetBasicTileEntity(i, j, out BasicTileEntity tileEntity)) {
+				return StatusText.Format(tileEntity.WaterFillPercentage);
+			}
+
+			// Note that it is possible for a map entry to be queried for a tile location that doesn't have a TileEntity anymore.
+			// This can happen in multiplayer when a world section hasn't been synced yet or when a players map hasn't been updated to match changes to the world.
+			// This code shows detecting those situations, but the basic lesson is don't assume a TileEntity will always be present.
+			/*
+			Point16 topLeft = TileObjectData.TopLeft(i, j);
+			int index = ModContent.GetInstance<BasicTileEntity>().Find(topLeft.X, topLeft.Y);
+			if (index == -1) {
+				if (!Main.sectionManager.TileLoaded(topLeft.X, topLeft.Y)) {
+					return $"{name}: World section not loaded yet";
+				}
+				return $"{name}: No TileEntity found at coordinate";
+			}
+			*/
+
+			return name;
+		}
+
+		public override void SetDrawPositions(int i, int j, ref int width, ref int offsetY, ref int height, ref short tileFrameX, ref short tileFrameY) {
+			if (TryGetBasicTileEntity(i, j, out BasicTileEntity tileEntity)) {
+				tileFrameY = (short)(tileFrameY + (tileEntity.WaterFillStage * 38));
+
+				// We can uncomment this code to spawn dust at the tile entity position for debugging purposes. Some developer mods also have tools to visualize tile entities.
+				/*
+				if (TileObjectData.IsTopLeft(i, j)) {
+					Dust.QuickDust(tileEntity.Position.X, tileEntity.Position.Y, Color.Green);
+				}
+				*/
+			}
+		}
+
+		public override void MouseOver(int i, int j) {
+			if (TryGetBasicTileEntity(i, j, out BasicTileEntity tileEntity)) {
+				Player player = Main.LocalPlayer;
+				player.noThrow = 2;
+				player.cursorItemIconEnabled = true;
+				player.cursorItemIconID = -1;
+				player.cursorItemIconText = StatusText.Format(tileEntity.WaterFillPercentage);
+			}
+		}
+
+		public override bool RightClick(int i, int j) {
+			if(TryGetBasicTileEntity(i, j, out BasicTileEntity tileEntity)) {
+				if(tileEntity.WaterFillPercentage == 100) {
+					Main.LocalPlayer.QuickSpawnItem(Main.LocalPlayer.GetSource_TileInteraction(i, j), ItemID.WaterBucket);
```

Comment:
Should the item be spawned by the server instead? If multiple players try and get water at the same time (within latency) they'll dupe it. Not the best example to set. What do item frames do?

Reply:
Tile has been updated to follow vanilla conventions, which is to do the item dropping through KillTile. This also allows the player to "mine" the tile to obtain the dropped item now. This also prevents any potential desync duplication.